File: backend/app/schemas/dual_coverage.py
"""Wire shapes for dual-coverage cases, opportunities and decisions."""
from __future__ import annotations

from typing import Literal, Self

from pydantic import BaseModel, Field, model_validator


class DualPartyOut(BaseModel):
    employee_id: str | None = None
    staff_id: str
    employee_name: str | None = None
    dependant_id: str | None = None
    relationship: str | None = None
    covered: bool = False
    covered_products: list[str] = Field(default_factory=list)
    # A dependant row whose sponsor is not an active employee, or that is not
    # linked at all: evidence the duplicate exists, never a live coverage line.
    unlinked: bool = False


class DualDecisionOut(BaseModel):
    decision: str
    carried_by_employee_id: str | None = None
    carried_by_staff_id: str | None = None
    note: str | None = None
    decided_by: str | None = None
    decided_at: str | None = None
    # True when the family composition changed after the decision was taken. The
    # case re-surfaces — the failure mode is always "ask again", never "silently
    # resolved".
    stale: bool = False


class DualCaseOut(BaseModel):
    subject_key: str
    name: str
    dob: str | None = None
    nric_masked: str | None = None
    relationship: str | None = None
    # "nric" is certain; "name_dob" is strong but occasionally wrong, which is
    # why `not_a_match` is a first-class decision.
    match_tier: Literal["nric", "name_dob"]
    flags: list[str] = Field(default_factory=list)
    parties: list[DualPartyOut] = Field(default_factory=list)
    # Products BOTH sides cover — the intersection, so this is real double
    # exposure rather than two unrelated covers.
    overlapping_products: list[str] = Field(default_factory=list)
    severity: Literal["warn", "info"]
    decision: DualDecisionOut | None = None


class DualLifeRefOut(BaseModel):
    """One dependant ROW's membership in a shared life.

    What a roster table needs to mark the row and NAME both employees, without
    carrying the case's decision workflow. Keyed by dependant row rather than by
    life because that is what a table has in hand.
    """

    dependant_id: str
    subject_key: str
    severity: Literal["warn", "info"]
    # Decided, and the decision still describes this family.
    resolved: bool
    # Every side of the life, in case order — the table names them all rather
    # than saying "also somewhere else".
    parties: list[DualPartyOut] = Field(default_factory=list)


class DualOpportunityOut(BaseModel):
    subject_key: str
    employees: list[DualPartyOut]
    child_name: str
    child_dob: str | None = None
    listed_under_staff_id: str
    other_staff_id: str
    decision: DualDecisionOut | None = None


class DualCoverageOut(BaseModel):
    """Counts are always exact; the lists are capped (`preview_cap`)."""

    unresolved_cases: int
    total_cases: int
    total_opportunities: int
    cases: list[DualCaseOut]
    opportunities: list[DualOpportunityOut]
    preview_cap: int
    # Deliberately NOT capped, unlike `cases`. This drives a per-row marker on a
    # PAGINATED table, so a cap would silently leave rows on later pages
    # unmarked — the one failure this column exists to prevent. Each entry is
    # small, and the list is bounded by the duplicates a roster actually has.
    lives: list[DualLifeRefOut] = Field(default_factory=list)


class DualDecisionIn(BaseModel):
    subject_key: str = Field(min_length=1, max_length=64)
    subject_kind: Literal["life", "couple"] = "life"
    decision: Literal["carried_by", "intentional_both", "not_a_match", "dismissed"] = (
        "carried_by"
    )
    carried_by_employee_id: str | None = None
    note: str | None = Field(default=None, max_length=2000)

    @model_validator(mode="after")
    def _carried_by_needs_an_employee(self) -> Self:
        # "Carried by nobody" is not a decision — it would clear the flag while
        # recording nothing about who keeps the life.
        if self.decision == "carried_by" and not self.carried_by_employee_id:
            raise ValueError("carried_by requires carried_by_employee_id.")
        return self

File: frontend/src/api/dualCoverage.ts
/**
 * Dual coverage — lives insured twice under one company.
 *
 * Detection is computed server-side on every read; only the broker's DECISION is
 * stored. `unresolved_cases` counts CASES only — opportunities (married
 * colleagues whose child is listed once) are the normal state of such a family
 * and would bury the real duplicates if counted.
 */
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { api } from "@/api/client";
import { useSession } from "@/stores/session";

export interface DualParty {
  employee_id: string | null;
  staff_id: string;
  employee_name: string | null;
  dependant_id: string | null;
  relationship: string | null;
  covered: boolean;
  covered_products: string[];
  /** Sponsor is not an active employee, or the row is unlinked: evidence the
   *  duplicate exists, never a live coverage line. */
  unlinked: boolean;
}

export interface DualDecision {
  decision: "carried_by" | "intentional_both" | "not_a_match" | "dismissed";
  carried_by_employee_id: string | null;
  carried_by_staff_id: string | null;
  note: string | null;
  decided_by: string | null;
  /** The decider's name. `decided_by` is a uuid and must never be printed. */
  decided_by_name: string | null;
  decided_at: string | null;
  /** The family changed after the decision was taken, so it no longer describes
   *  the situation and the case is counted as unresolved again. */
  stale: boolean;
}

export interface DualCase {
  subject_key: string;
  name: string;
  dob: string | null;
  nric_masked: string | null;
  relationship: string | null;
  match_tier: "nric" | "name_dob";
  flags: ("listed_twice" | "employee_as_spouse")[];
  parties: DualParty[];
  overlapping_products: string[];
  severity: "warn" | "info";
  decision: DualDecision | null;
}

/** One dependant ROW's membership in a shared life — enough for a roster table
 *  to mark the row and name every employee the life reaches. Never capped
 *  server-side: the table is paginated, so a cap would leave later pages
 *  silently unmarked. */
export interface DualLifeRef {
  dependant_id: string;
  subject_key: string;
  severity: "warn" | "info";
  resolved: boolean;
  parties: DualParty[];
}

export interface DualOpportunity {
  subject_key: string;
  employees: DualParty[];
  child_name: string;
  child_dob: string | null;
  listed_under_staff_id: string;
  other_staff_id: string;
  decision: DualDecision | null;
}

export interface DualCoverage {
  unresolved_cases: number;
  total_cases: number;
  total_opportunities: number;
  cases: DualCase[];
  opportunities: DualOpportunity[];
  preview_cap: number;
  lives: DualLifeRef[];
}

/** dependant row id → the shared life it belongs to. */
export function livesByDependant(data?: DualCoverage): Map<string, DualLifeRef> {
  return new Map((data?.lives ?? []).map((l) => [l.dependant_id, l]));
}

/**
 * `focus` names one case the response must contain whatever `preview_cap` does.
 * The table marks its rows from `lives`, which is uncapped, so without it a row
 * past the cap opened the sheet onto a case that wasn't in `cases` — reported as
 * "Nothing open for this person" on the row that had just been clicked.
 */
export function useDualCoverage(
  policyYearId: string | undefined,
  focus?: string | null,
) {
  const clientId = useSession((s) => s.activeClientId);
  return useQuery({
    queryKey: ["dependants", "dual-coverage", clientId, policyYearId, focus ?? null],
    queryFn: () =>
      api.get<DualCoverage>(
        `/policy-years/${policyYearId}/dual-coverage${
          focus ? `?focus=${encodeURIComponent(focus)}` : ""
        }`,
      ),
    enabled: !!policyYearId,
  });
}

export function useRecordDualDecision(policyYearId: string | undefined) {
  const qc = useQueryClient();
  const clientId = useSession((s) => s.activeClientId);
  return useMutation({
    mutationFn: (body: {
      subject_key: string;
      subject_kind?: "life" | "couple";
      decision: DualDecision["decision"];
      carried_by_employee_id?: string | null;
      note?: string | null;
    }) =>
      api.post(`/policy-years/${policyYearId}/dual-coverage/decisions`, body),
    onSuccess: () =>
      qc.invalidateQueries({
        queryKey: ["dependants", "dual-coverage", clientId, policyYearId],
      }),
  });
}

/**
 * Drop or restore ONE side's cover for a life listed under two employees.
 *
 * This is the only control here that MOVES MONEY — the decision buttons record
 * who carries the life, this changes who pays for them. Both rows stay on the
 * roster either way. Invalidates the coverage-bearing queries too, since a
 * dropped side changes the benefit statement and the member's flex tier.
 */
export function useSetDualCover(policyYearId: string | undefined) {
  const qc = useQueryClient();
  const clientId = useSession((s) => s.activeClientId);
  return useMutation({
    mutationFn: ({
      dependantId,
      covered,
    }: {
      dependantId: string;
      covered: boolean;
    }) =>
      api.put<{ covered: boolean; products_changed: string[] }>(
        `/policy-years/${policyYearId}/dual-coverage/dependants/${dependantId}/cover`,
        { covered },
      ),
    onSuccess: () => {
      qc.invalidateQueries({
        queryKey: ["dependants", "dual-coverage", clientId, policyYearId],
      });
      for (const key of [
        "benefit-statement",
        "plan-overrides",
        "flex-membership",
        "flex-coverage",
        "employees",
      ]) {
        qc.invalidateQueries({ queryKey: [key] });
      }
    },
  });
}

export function useReopenDualDecision(policyYearId: string | undefined) {
  const qc = useQueryClient();
  const clientId = useSession((s) => s.activeClientId);
  return useMutation({
    mutationFn: (subjectKey: string) =>
      api.delete(
        `/policy-years/${policyYearId}/dual-coverage/decisions/${subjectKey}`,
      ),
    onSuccess: () =>
      qc.invalidateQueries({
        queryKey: ["dependants", "dual-coverage", clientId, policyYearId],
      }),
  });
}
